### ISSUE TITLE
Improved some error messagge in reflect.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3259]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Wolf reported|
|-|-|
|Trac ID|trac#3259|
|Type|enhancement|
|Created|2008-05-30 20:52:37Z|
|Branch|https://github.com/twisted/twisted/tree/namedany-errors-3259|

I added some checks in the namedAny method in twisted.python.reflect in order to display better error messagges. When the object passed didn't exist an "Empty module name" error was raised because the function sent an empty string to _importAndCheckStack. I added an if/else to prevent this and output the name of the non-existing object. Another if/else tries to understand if the name passed was supposed to be a module or another kind of object. If an empty string or a name that ends with '.' are passed, an error is now raised without parsing the argument.
I tried it with trial, even if this function is probably used somewhere else:

"trial nonexistingmodule" now outputs:
```
exceptions.ValueError: No module named nonexistingmodule.py
```
instead of:
```
   return __import__(importName)
exceptions.ValueError: Empty module name
```

"trial x.y.z" now outputs:
```
exceptions.ValueError: exceptions.ValueError: "x.y.z" is not a valid 
dot-separated list of Python objects (packages, modules, classes or any 
other Python object which has attributes)
```
instead of:
```
   return __import__(importName)
exceptions.ValueError: Empty module name
```

"trial twisted.test." now outputs:
```
exceptions.ValueError: Invalid name twisted.test. (it shouldn't end with a '.')
```
instead of:
```
exceptions.AttributeError: 'module' object has no attribute ''
```

I also added a test to test_reflect.py in order to test 'twisted.test.'

Attachments:

* [ticket3259.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ed8/ed86f246f734646d674dd346e5bb88f9d2efbabf/797809c2058e9bc8cf72188cab007ad1cf8569a7.patch) (6565 bytes) - added by Wolf on 2008-06-02 17:10:04Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3259 3259
type__enhancement enhancement
reporter__Wolf Wolf
priority__normal normal
milestone__ 
branch__branches_namedany_errors_3259 branches/namedany-errors-3259
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__reflect_namedAny reflect namedAny
time__1212180757000000 1212180757000000
changetime__1212490829000000 1212490829000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve set owner to Wolf|
|-|-|

Thanks for your contribution, but it's missing tests. You need at least 4 new tests, as 4 new exceptions are introduced. Also, you should check the message of the exception (assertRaises returns the exception).

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Wolf removed owner|
|-|-|

Added a few more tests and the check for the messages. I also raise an error if the name starts with a '.' because the __import__ in _importAndCheckStack will fail to load it (it will raise an 'Empty module name'). All the new exceptions added replace the generic exception raised before, so instead of adding 5 more test I changed the old ones that already covered some of the cases.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to Wolf|
|-|-|

Patch is incompatible with Python 2.3, a version of Python which Twisted supports.

Also, the test that does `namedAny('dummy.py')` will fail if someone has a module named `dummy`.

I'd definitely prefer different exception types for each error case, rather than just different strings stashed inside `ValueError` instances.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Wolf removed owner|
|-|-|

Changed err.message to str(err) in order to be compatible with py 2.3.

Renamed dummy.py with nosuchmoduleintheworld.py.

Created 3 new exceptions: InvalidName(ValueError) used when the name is not valid, ModuleNotFound(InvalidName) used when the name is not an existing module and ObjectNotFound(InvalidName) used when the name is not an existing object.

Fixed the tests in order to work with the new exceptions and modified a couple of docstrings in order to include the new Exceptions raised (I couldn't test the epytext syntax, I hope it's correct)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [23835]) Branching to 'namedany-errors-3259'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [23836]) Apply ticket43259.patch

refs [#3259](https://github.com/twisted/twisted/issues/3259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [23838]) clean up documentation language a little bit

refs [#3259](https://github.com/twisted/twisted/issues/3259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [23839]) get rid of .py/.pyc handling in namedAny; add internal '..' handling

refs [#3259](https://github.com/twisted/twisted/issues/3259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [23840]) ummm.  change the text expected.

maybe trial can handle these new exception types specially now

refs [#3259](https://github.com/twisted/twisted/issues/3259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [23841]) add an `__all__` and fix some other trivial stuff

refs [#3259](https://github.com/twisted/twisted/issues/3259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [23842]) copyright, module docstring, remove unused import

refs [#3259](https://github.com/twisted/twisted/issues/3259)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Some changes, as evidenced above.  Probably in good shape now.  Someone else should take a look.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @exarkun|
|-|-|

That's great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [23850]) Merge namedany-errors-3259

Author: wolf, exarkun
Reviewer: exarkun, therve
Fixes: [#3259](https://github.com/twisted/twisted/issues/3259)

Provide a different exception type for each of several different
failures which namedAny can have.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
